src/aws/flb_s3_local_buffer.c

```diff
@@ -414,3 +414,110 @@ int flb_remove_chunk_files(struct flb_local_chunk *c)
     }
     return ret;
 }
+
+/* Generic replace function for strings. */
+char* replace_uri_tokens(const char* original_string, const char* current_word, const char* new_word) 
+{ 
+    char *result; 
+    int i = 0;
+    int count = 0; 
+    int new_word_len = strlen(new_word); 
+    int old_word_len = strlen(current_word); 
+  
+    for (i = 0; original_string[i] != '\0'; i++) { 
+        if (strstr(&original_string[i], current_word) == &original_string[i]) { 
+            count++; 
+            i += old_word_len - 1; 
+        } 
+    }
+
+    result = flb_sds_create_size(i + count * (new_word_len - old_word_len) + 1);
+    if (!result) {
+        flb_errno();
+        return NULL;
+    }
+
+    i = 0; 
+    while (*original_string) { 
+        if (strstr(original_string, current_word) == original_string) { 
+            strcpy(&result[i], new_word); 
+            i += new_word_len; 
+            original_string += old_word_len; 
+        } 
+        else
+            result[i++] = *original_string++; 
+    } 
+
+    result[i] = '\0'; 
+    return result; 
+} 
+  
+//* Constructs S3 object key as per the format. */
+flb_sds_t get_s3_key(const char *format, time_t time, char *tag, char *tag_delimiter)
+{
+    int i = 0;
+    int max_key_len = (strlen(format) + strlen(tag) + 1);
+    char tag_idx[9];
+    char open[2] = "[";
+    char close[2] = "]";
+    char *tag_token = NULL;
+    char tag_descriptor[10] = "$TAG";
+    char datestamp[max_key_len];
+    flb_sds_t s3_key = NULL;
+    flb_sds_t tmp_key = NULL;
+    char *tmp_tag = NULL;
+    struct tm *gmt = NULL;
+    
+    tmp_tag = flb_malloc(strlen(tag) + 1);
+    s3_key = flb_sds_create_size(max_key_len);
+    if (!s3_key) {
+        flb_errno();
+        return NULL;
+    }
+    strcpy(s3_key, format);
+    strcpy(tmp_tag, tag);
```

Comment:
I wouldn't do either of these creations this way.... instead do:

s3_key = flb_sds_create_len(format, strlen(format));
// check...
// make tmp_tag an sds?
tmp_tag = flb_sds_create_len(tag, strlen(tag));

Reply:
Also, why do you even need to copy the tag and the format anyway? Later on you set s3_key to be the output of `replace_uri_tokens` which allocates a new SDS string. Which means the guess for the size of it doesn't matter... since you're overwriting the original. Which begs the question- why even create s3_key as a copy of format here?

Reply:
If you want we can go over this code together... you can explain to me how you are thinking about it and what you think the best strategy is for allocating the strings and I can try to figure out how I would do it.

Reply:
Here is the logic in a gist:
1) split the tag on the delimiter/(s). This will return a pointer to the first token found in the string.
2) search for occurrences of $TAG[*], 
       if found replace it in the string with the respective part of tag.
      note:  at each stage the replace uri takes (current string, characters to be replaced, new characters)
3) now that all occurrences of $TAG[*] are taken care of, we need to replace the case where $TAG has occurred.
    eg: logs/$TAG[2]/$TAG/%Y
4) last step is to format the time and return the result.

The reason we need a copy of the tag
    when you do strtok to split the tag, we get a pointer to the first token. By the time we have reached step 3, the tag 
    with its delimiters is no longer available in its entirety for us to replace in a simple way. This is the reason, I created 
    the copy.

The reason we need a copy for format.
    Ideally we should not modify the function parameter format. Secondly, in the while loop at step 2, it is important 
    that we send the most updated string to the function. We cannot do s3_key = replace_uri_tokens(s3_key, tmp, 
    tag_token); due to memory leak, and neither can we send format here because the tags replaced won't be 
    available in format string.